src/app/clusters/time-synchronization-server/time-synchronization-server.cpp

```diff
@@ -616,30 +606,42 @@ CHIP_ERROR TimeSynchronizationServer::SetTimeZone(const DataModel::DecodableList
     }
     if (CHIP_NO_ERROR != newTzL.GetStatus())
     {
-        ReturnErrorOnFailure(LoadTimeZone());
-        return CHIP_IM_GLOBAL_STATUS(InvalidCommand);
+        ExitNow(err = CHIP_IM_GLOBAL_STATUS(InvalidCommand));
     }
 
     mTimeZoneObj.validSize = i;
 
-    if (lastTzState != TimeState::kInvalid && TimeState::kInvalid != UpdateTimeZoneState())
+    currentTzState = UpdateTimeZoneState();
+    if (lastTzState == TimeState::kActive && TimeState::kActive == currentTzState)
     {
         bool emit       = false;
         const auto & tz = GetTimeZone()[0].timeZone;
         if (tz.offset != lastTz.offset)
         {
             emit = true;
         }
-        if ((tz.name.HasValue() && lastTz.name.HasValue()) && !(tz.name.Value().data_equal(lastTz.name.Value())))
+        if (tz.name.HasValue() != lastTz.name.HasValue() ||
+            ((tz.name.HasValue() && lastTz.name.HasValue()) && !(tz.name.Value().data_equal(lastTz.name.Value()))))
         {
             emit = true;
         }
         if (emit)
         {
-            mEventFlag = TimeSyncEventFlag::kTimeZoneStatus;
+            emitTimeZoneStatusEvent(GetDelegate()->GetEndpoint());
         }
     }
-    return mTimeSyncDataProvider.StoreTimeZone(GetTimeZone());
+
+exit:
+    if (err != CHIP_NO_ERROR)
+    {
+        // do not propagate if an error occurs here because it is not the user relevant error
```

Comment:
```suggestion
        // Restore our state to the last known-good one.
        //
        // Do not propagate failure from LoadTimeZone if an error occurs here because it is not the user relevant error
```